Vaadin 8 seems to require bean validation library, but this is not declared in documentation

With Vaadin 8.0.0.beta1 and BeanBinder you'll get an exception if bean validation library is not present. In previous versions Vaadin versions this was not required.

Either the functionality should be similar as before (just gracefully continue with limited featureset) or documentation and archetypes should be fixed.

See https://github.com/mstahv/v-leaflet/pull/144#issuecomment-270355264

For the future references :
- the project is outdated (BeanBinder doesn't no longer exist)
- there are references to 8.0-beta1 for compatibility module though everything else is 8.0-SNAPSHOT
- quite tricky to build (g-leaflet 1.4.0-SNAPSHOT dependency  is required though it's not available as a binary. One has to build it and install locally from sources)
- mistakes in the documentation about UI tests
- spend a lot of time to get it working and find a stacktrace (would be awesome to see the stacktrace in the ticket description from the very beginning).

Now the localized issue:

the root cause is : there is validation-api/1.0.0.GA/validation-api-1.0.0.GA.jar in the project's classpath. But there are no any JSR-303 provider available.
As a result : our code detects the JSR 303 API presence and thinks that bean validation can be used. Then BeanValidator class becomes available to validation and used for this. But it requires a validation provider available. And it fails at this point.
The stacktrace is below.

java.lang.ExceptionInInitializerError
	at com.vaadin.data.validator.BeanValidator.getJavaxBeanValidatorFactory(BeanValidator.java:145)
	at com.vaadin.data.validator.BeanValidator.getJavaxBeanValidator(BeanValidator.java:154)
	at com.vaadin.data.validator.BeanValidator.apply(BeanValidator.java:120)
	at com.vaadin.data.Binder$ValidatorAsConverter.convertToModel(Binder.java:989)
	at com.vaadin.data.Converter$2.lambda$convertToModel$5daf174e$1(Converter.java:167)
	at com.vaadin.data.SimpleResult.flatMap(SimpleResult.java:64)
	at com.vaadin.data.Converter$2.convertToModel(Converter.java:167)
	at com.vaadin.data.Binder$BindingImpl.doConversion(Binder.java:838)
	at com.vaadin.data.Binder$BindingImpl.doValidation(Binder.java:858)
	at com.vaadin.data.Binder$BindingImpl.access$1300(Binder.java:760)
	at com.vaadin.data.Binder.validateBindings(Binder.java:1596)
	at com.vaadin.data.Binder.doWriteIfValid(Binder.java:1451)
	at com.vaadin.data.Binder.writeBean(Binder.java:1406)
	at org.vaadin.addon.leaflet.demoandtestapp.JtsPointFieldTest.lambda$getTestComponent$166a13ad$1(JtsPointFieldTest.java:140)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.vaadin.event.ListenerMethod.receiveEvent(ListenerMethod.java:510)
	at com.vaadin.event.EventRouter.fireEvent(EventRouter.java:211)
	at com.vaadin.event.EventRouter.fireEvent(EventRouter.java:174)
	at com.vaadin.server.AbstractClientConnector.fireEvent(AbstractClientConnector.java:1028)
	at com.vaadin.ui.Button.fireClick(Button.java:369)
	at com.vaadin.ui.Button$1.click(Button.java:57)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.vaadin.server.ServerRpcManager.applyInvocation(ServerRpcManager.java:158)
	at com.vaadin.server.ServerRpcManager.applyInvocation(ServerRpcManager.java:119)
	at com.vaadin.server.communication.ServerRpcHandler.handleInvocation(ServerRpcHandler.java:443)
	at com.vaadin.server.communication.ServerRpcHandler.handleInvocations(ServerRpcHandler.java:415)
	at com.vaadin.server.communication.ServerRpcHandler.handleRpc(ServerRpcHandler.java:274)
	at com.vaadin.server.communication.UidlRequestHandler.synchronizedHandleRequest(UidlRequestHandler.java:90)
	at com.vaadin.server.SynchronizedRequestHandler.handleRequest(SynchronizedRequestHandler.java:41)
	at com.vaadin.server.VaadinService.handleRequest(VaadinService.java:1450)
	at com.vaadin.server.VaadinServlet.service(VaadinServlet.java:380)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:800)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:587)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:577)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:223)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1125)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1059)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
	at org.eclipse.jetty.server.Server.handle(Server.java:497)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:313)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:248)
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:540)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:626)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:546)
	at java.lang.Thread.run(Thread.java:745)
Caused by: javax.validation.ValidationException: Unable to find a default provider
	at javax.validation.Validation$GenericBootstrapImpl.configure(Validation.java:264)
	at javax.validation.Validation.buildDefaultValidatorFactory(Validation.java:111)
	at com.vaadin.data.validator.BeanValidator$LazyFactoryInitializer.getFactory(BeanValidator.java:190)
	at com.vaadin.data.validator.BeanValidator$LazyFactoryInitializer.<clinit>(BeanValidator.java:187)
	... 56 more

--- a/server/src/main/java/com/vaadin/data/util/BeanUtil.java
+++ b/server/src/main/java/com/vaadin/data/util/BeanUtil.java
@@ -20,6 +20,7 @@ import java.beans.IntrospectionException;
 import java.beans.Introspector;
 import java.beans.PropertyDescriptor;
 import java.io.Serializable;
+import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
 import java.util.ArrayList;
 import java.util.List;
@@ -229,14 +230,21 @@ public final class BeanUtil implements Serializable {
 
         private static boolean isAvailable() {
             try {
-                Class.forName("javax.validation.Validation");
+                Class<?> clazz = Class.forName("javax.validation.Validation");
+                Method method = clazz.getMethod("buildDefaultValidatorFactory");
+                method.invoke(null);
                 return true;
-            } catch (ClassNotFoundException e) {
+            } catch (ClassNotFoundException | NoSuchMethodException
+                    | InvocationTargetException e) {
                 Logger.getLogger(BeanValidator.class.getName())
                         .fine("A JSR-303 bean validation implementation not found on the classpath. "
                                 + BeanValidator.class.getSimpleName()
                                 + " cannot be used.");
                 return false;
+            } catch (IllegalAccessException | IllegalArgumentException e) {
+                throw new RuntimeException(
+                        "Unable to invoke javax.validation.Validation.buildDefaultValidatorFactory()",
+                        e);
             }
         }
 

--- a/test/bean-api-validation/pom.xml
+++ b/test/bean-api-validation/pom.xml
@@ -1,0 +1,41 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>com.vaadin</groupId>
+        <artifactId>vaadin-test</artifactId>
+        <version>8.0-SNAPSHOT</version>
+    </parent>
+    <artifactId>vaadin-test-bean-api-validation</artifactId>
+    <packaging>jar</packaging>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.eclipse.jetty</groupId>
+                <artifactId>jetty-maven-plugin</artifactId>
+                <version>${jetty.version}</version>
+                <configuration>
+                    <scanIntervalSeconds>-1</scanIntervalSeconds>
+                    <stopPort>8081</stopPort>
+                    <stopWait>5</stopWait>
+                    <stopKey>foo</stopKey>
+                </configuration>
+                <executions>
+                    <!-- start and stop jetty (running our app) when running 
+                        integration tests -->
+                    <execution>
+                        <id>start-jetty</id>
+                        <phase></phase>
+                    </execution>
+                    <execution>
+                        <id>stop-jetty</id>
+                        <phase></phase>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
+
+</project>

--- a/test/bean-api-validation/src/test/java/com/vaadin/data/BeanValidationTest.java
+++ b/test/bean-api-validation/src/test/java/com/vaadin/data/BeanValidationTest.java
@@ -1,0 +1,63 @@
+/*
+ * Copyright 2000-2016 Vaadin Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.vaadin.data;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import com.vaadin.ui.TextField;
+
+/**
+ * @author Vaadin Ltd
+ *
+ */
+public class BeanValidationTest {
+
+    public static class Bean {
+
+        private String property;
+
+        public String getProperty() {
+            return property;
+        }
+
+        public void setProperty(String property) {
+            this.property = property;
+        }
+
+    }
+
+    @Test
+    public void binderWorksWithoutBeanValidationLib() {
+        try {
+            Class.forName("javax.validation.Validation");
+            Assert.fail();
+        } catch (ClassNotFoundException ignored) {
+        }
+
+        Binder<Bean> binder = new Binder<>(Bean.class);
+
+        TextField field = new TextField();
+        binder.forField(field).bind("property");
+
+        Bean bean = new Bean();
+        binder.setBean(bean);
+
+        field.setValue("foo");
+        Assert.assertEquals(field.getValue(), bean.getProperty());
+
+    }
+}

--- a/test/bean-impl-validation/pom.xml
+++ b/test/bean-impl-validation/pom.xml
@@ -1,0 +1,50 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>com.vaadin</groupId>
+        <artifactId>vaadin-test</artifactId>
+        <version>8.0-SNAPSHOT</version>
+    </parent>
+    <artifactId>vaadin-test-bean-impl-validation</artifactId>
+    <packaging>jar</packaging>
+
+
+    <dependencies>
+        <dependency>
+            <groupId>javax.validation</groupId>
+            <artifactId>validation-api</artifactId>
+            <version>1.1.0.Final</version>
+        </dependency>
+    </dependencies>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.eclipse.jetty</groupId>
+                <artifactId>jetty-maven-plugin</artifactId>
+                <version>${jetty.version}</version>
+                <configuration>
+                    <scanIntervalSeconds>-1</scanIntervalSeconds>
+                    <stopPort>8081</stopPort>
+                    <stopWait>5</stopWait>
+                    <stopKey>foo</stopKey>
+                </configuration>
+                <executions>
+                    <!-- start and stop jetty (running our app) when running 
+                        integration tests -->
+                    <execution>
+                        <id>start-jetty</id>
+                        <phase></phase>
+                    </execution>
+                    <execution>
+                        <id>stop-jetty</id>
+                        <phase></phase>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
+
+</project>

--- a/test/bean-impl-validation/src/test/java/com/vaadin/data/BeanValidationTest.java
+++ b/test/bean-impl-validation/src/test/java/com/vaadin/data/BeanValidationTest.java
@@ -1,0 +1,63 @@
+/*
+ * Copyright 2000-2016 Vaadin Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.vaadin.data;
+
+import javax.validation.Validation;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import com.vaadin.ui.TextField;
+
+/**
+ * @author Vaadin Ltd
+ *
+ */
+public class BeanValidationTest {
+
+    public static class Bean {
+
+        private String property;
+
+        public String getProperty() {
+            return property;
+        }
+
+        public void setProperty(String property) {
+            this.property = property;
+        }
+
+    }
+
+    @Test
+    public void binderWorksWithoutBeanValidationImpl() {
+        // Just to make sure that it's available at the compilation time and in
+        // runtime
+        Assert.assertNotNull(Validation.class);
+
+        Binder<Bean> binder = new Binder<>(Bean.class);
+
+        TextField field = new TextField();
+        binder.forField(field).bind("property");
+
+        Bean bean = new Bean();
+        binder.setBean(bean);
+
+        field.setValue("foo");
+        Assert.assertEquals(field.getValue(), bean.getProperty());
+
+    }
+}

--- a/test/pom.xml
+++ b/test/pom.xml
@@ -99,6 +99,8 @@
         <module>vaadinservletconfiguration-widget-set</module>
         <module>spring-boot</module>
         <module>cdi</module>
+        <module>bean-api-validation</module>
+        <module>bean-impl-validation</module>
     </modules>
 
     <build>

